aten/src/THC/THCTensorRandom.cuh

```diff
@@ -94,15 +94,19 @@ template <typename T>
 __global__ void renormRowsL1(T* dist, long rows, long cols) {
   extern __shared__  unsigned char my_smem[];
   T *smem = reinterpret_cast<T *>(my_smem);
-
+  T zero = ScalarConvert<int, T>::to(0);
+  T val;
   for (int64_t row = blockIdx.x; row < rows; row += gridDim.x) {
     T sum = ScalarConvert<int, T>::to(0);
     for (int64_t col = threadIdx.x; col < cols; col += blockDim.x) {
-      sum = THCNumerics<T>::add(sum, dist[row * cols + col]);
+      val = dist[row * cols + col];
+      assert(THCNumerics<T>::ge(val, zero));
+      sum = THCNumerics<T>::add(sum, val);
     }
 
-    sum = reduceBlock(smem, blockDim.x, sum, ReduceAdd<T, T>(), ScalarConvert<int, T>::to(0));
+    sum = reduceBlock(smem, blockDim.x, sum, ReduceAdd<T, T>(), zero);
     if (threadIdx.x == 0) {
+      assert(THCNumerics<T>::gt(sum, zero));
```

Comment:
I'm not sure we need an assertion here. You are checking in the above loop that there are no negative values, so there should be no case that leads to a negative value from the reduction.

Reply:
They can be all zeros.

Reply:
Ah you are right, my b.